optuna/notebooks/jupytercon_demo/RAPIDS_xfeat_Optuna-GPU.ipynb

```diff
@@ -0,0 +1,1803 @@
+    "\n",
+    "Now that we have some new features, how do we know they are relevant for the task or represent anything meaningful? We use the feature selection process to do this. This helps in selection of a subset of features that are most informative. This helps in simplifying the problem and ensures that we aren't overloading the system with unimportant features.\n",
+    "\n",
+    "Optuna provides a way to choose a selector which accepts a Pipeline object from xfeat. You can see in the `feature_selection` function we define a `Pipeline` that takes in an Explorer and a Selection Algorithm (`ChiSquareKBest`). We pass this to an Optuna Study object, along with an Objective function. Chi squared tests are used to test the independence of two events.\n",
+    "\n",
+    "For Feature Selection, we aim to select features, which are highly dependent on the response. This way, we can get features that will best determine the outcome.\n",
+    "\n",
+    "#### Objective Function\n",
+    "The objective function will be the one we optimize in Optuna Study. Objective function tries out different values for the parameters that we are tuning and saving the results in `study.trials_dataframes()`.\n",
+    "\n",
+    "Let's define the objective function for this HPO task by making use of the `train_and_eval()`. You can see that we simply choose a value for the parameters and call the train_and_eval method, making Optuna very easy to use in an existing workflow.\n",
+    "\n",
+    "The objective remains constant over different samplers, which are built-in options in Optuna to enable the selection of different sampling algorithms that optuna provides. Some of the available ones include - GridSampler, RandomSampler, TPESampler, etc. We'll use TPESampler for this demo, but feel free to try different samplers to notice the changes in performance. Tree-structured Parzen Estimator or TPE is based on independent sampling. It fits a Gaussian Mixture Model for each parameter in a trial `l(x)` and another GMM `g(x)` for other parameters. And it optimizes the `l(x)/g(x)` ratio to yield the best parameters.\n",
+    "\n",
+    "### HPO Trials and Study\n",
+    "Optuna uses [study](https://optuna.readthedocs.io/en/stable/reference/study.html) and [trials](https://optuna.readthedocs.io/en/stable/reference/trial.html) to keep track of the HPO experiments. Put simply, a trial is a single call of the objective function while a set of trials make up a study. We will pick the optimal performing trial from a study to get the best parameters that were used in that run."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 5,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def train_and_eval(df, penalty='l2', C=1.0, l1_ratio=None, fit_intercept=True, selector=None, return_model=False):\n",
+    "    # Splitting data and prepping for selector fit\n",
+    "    X_train,  X_test, y_train, y_test = train_test_split(df,\n",
```

Comment:
How is the test set passed through here? We want to make sure that the test set held out in feature engineering (esp target encoding) is not used for training, but I don't see how that works here? Similarly, the inner eval loop seems to split the data differently? Maybe I'm just misinterpreting. though.

Reply:
I'm not sure I understand why the test set in Feature Engineering should not be used here? With feature engineering - I split the data and perform the encoding. Once the data is changed, I'm concatenating the dataframes to continue the rest of the workflow.

Which loop are you referring to? I've set the random_state for all splits to ensure that the results are reproducible with the splits.